### PR TITLE
bazel: various improvements to generation

### PR DESCRIPTION
**Pull request by ajwerner**
This adds targets for `:stringer`, `:gomock`, `:execgen`, `:optgen`, `:misc`, and `:docs` to `//pkg/gen` and then some aliases `:code` and `:gen`. This PR then leverages these targets to do all code generation.

### CODE REVIEW

File: pkg/gen/gen.bzl [hunk 1]
```diff
@@ -20,82 +84,184 @@ def _go_proto_srcs_impl(ctx):
...
+   "_srcs": attr.label_list(providers = [GoSource], default=PROTOBUF_SRCS),
+  },
 )
 
-_cp_file_cmd_fmt = """
+# This rule is the default rule to build construct the input to _hoist_files
+# for srcs which have a path in the sandbox that is parallel to where those
+# files should end up in the repo.
+def _no_prefix_impl(ctx):
+  files = [f for di in ctx.attr.srcs for f in di[DefaultInfo].files.to_list()]
```
Comment (irfansharif) [inline pkg/gen/gen.bzl:109]: Out of curiosity: how'd you work through the starlark API here? Is there an IDE like thing? Looking at docs? Using python reflection to figure out what fields/methods are available + printf-ing your way through?

File: pkg/gen/genbzl/main.go [hunk 2]
```diff
@@ -46,46 +46,90 @@ func main() {
...
+	{
+		filename: "docs.bzl", variable: "DOCS_SRCS",
+		query: `
+kind("generated file", //docs/...:*)
+  - labels("outs", //docs/generated/sql/bnf:svg)`,
+	},
+	{
+		filename: "excluded.bzl", variable: "EXCLUDED_SRCS",
+		query: `
+let all = kind("generated file", //...:*)
```
Comment (irfansharif) [inline pkg/gen/genbzl/main.go:100]: TIL how expressive bazel queries can be.

**Review by irfansharif — APPROVED** [review]
Very cool.

**Review by ajwerner — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76355#-:-MvesVDb9XsVoVnoxE4-:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @irfansharif)

___
*[pkg/gen/gen.bzl, line 1 at r29](https://reviewable.io/reviews/cockroachdb/cockroach/76355#-MvWnoeuFBkZNZ16StSZ-r29-1:-MvesIZu2OmbHOZrO4md:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/851b5b72029f216f11b59a2e12806252643c482d/pkg/gen/gen.bzl#L1)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

[nit] Empty line.
</blockquote></details>

Done.
___
*[pkg/gen/gen.bzl, line 39 at r29](https://reviewable.io/reviews/cockroachdb/cockroach/76355#-MvWnoeuFBkZNZ16StSZ-r29-39:-MvesIqkCvi-FsFOmf4A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/851b5b72029f216f11b59a2e12806252643c482d/pkg/gen/gen.bzl#L39)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

s/th/the
</blockquote></details>

Done.
___
*[pkg/gen/gen.bzl, line 109 at r29](https://reviewable.io/reviews/cockroachdb/cockroach/76355#-MvWnoeuFBkZNZ16StSZ-r29-109:-Mve9asBCpeO5NQjNgg0:by6xfjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/851b5b72029f216f11b59a2e12806252643c482d/pkg/gen/gen.bzl#L109)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Out of curiosity: how'd you work through the starlark API here? Is there an IDE like thing? Looking at docs? Using python reflection to figure out what fields/methods are available + printf-ing your way through?
</blockquote></details>

There were definitely some moments of sprinkling in `print` statements and poking around. `dir` got used once or twice. Most of it though is just reading the reference. The docs are pretty reasonable https://docs.bazel.build/versions/main/skylark/lib/DefaultInfo.html. 
___
*[pkg/sql/opt/xform/BUILD.bazel, line 117 at r29](https://reviewable.io/reviews/cockroachdb/cockroach/76355#-MvXR1wHsddzat40nG-q-r29-117:-MveA4ua6VBx3iwsX8hD:bepi4og) ([raw file](https://github.com/cockroachdb/cockroach/blob/851b5b72029f216f11b59a2e12806252643c482d/pkg/sql/opt/xform/BUILD.bazel#L117)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Do you think it'd be worth wrapping `genrule` with a macro of our own that uses this set of visibility rules out of the box? Ditto for `gomock`.
</blockquote></details>

I'll do that in a follow-up. I'd like to get this merged.

**Comment by ajwerner** [comment]
TFTR!
🙏  I don't mess up anybody's workflow too badly

bors r=irfansharif

**Pull request merged**